tests/integration/src/test/java/org/mule/context/notification/ProcessorsTraceTestCase.java

```diff
@@ -0,0 +1,516 @@
+
+        assertThat(ProcessorsTraceAsserter.processorsTraceToAssert, not(nullValue()));
+
+        assertThat(ProcessorsTraceAsserter.processorsTraceToAssert, hasExecutedProcessors(
+                "/flowDynamicWithScatterGatherChain/processors/0",
+                "/flowDynamicWithScatterGatherChain/processors/0/0/0",
+                "/flowDynamicWithScatterGatherChain/processors/0/1",
+                "/flowDynamicWithScatterGatherChain/processors/0/1/0",
+                "/flow/processors/0"));
+    }
+
+    @Test
+    public void subFlowDynamicWithScatterGatherChain() throws Exception
+    {
+        muleContext.getClient().send("vm://in-subFlowDynamicWithScatterGatherChain", new DefaultMuleMessage("payload", muleContext));
+
+        assertThat(ProcessorsTraceAsserter.processorsTraceToAssert, not(nullValue()));
+        
+        assertThat(ProcessorsTraceAsserter.processorsTraceToAssert, hasExecutedProcessors(
+                "/subFlowDynamicWithScatterGatherChain/processors/0",
+                "/subFlowDynamicWithScatterGatherChain/processors/0/0/0",
+                "/subFlowDynamicWithScatterGatherChain/processors/0/1",
+                "/subFlowDynamicWithScatterGatherChain/processors/0/1/0",
+                "/subFlowDynamicWithScatterGatherChain/processors/0/1/0/subFlow/subprocessors/0"));
+    }
+
+    private Matcher<ProcessorsTrace> hasExecutedProcessors(final String... expectedProcessors)
+    {
+        return new TypeSafeMatcher<ProcessorsTrace>()
+        {
+            private List<Matcher> failed = new ArrayList<>();
+            private List<String> actualProcessors;
+            
+            @Override
+            protected boolean matchesSafely(ProcessorsTrace processorsTrace)
+            {
+                Matcher<Collection<? extends Object>> sizeMatcher = hasSize(expectedProcessors.length);
+                if (!sizeMatcher.matches(processorsTrace.getExecutedProcessors()))
+                {
+                    failed.add(sizeMatcher);
+                }
+
+                int i = 0;
+                for (String expectedProcessor : expectedProcessors)
+                {
+                    Matcher processorItemMatcher = startsWith(expectedProcessor + " @");
+                    if (!processorItemMatcher.matches(processorsTrace.getExecutedProcessors().get(i)))
+                    {
+                        failed.add(processorItemMatcher);
+                    }
+                    ++i;
+                }
+
+                return failed.isEmpty();
+            }
+            
+            @Override
+            public void describeTo(Description description)
+            {
+                description.appendText(Arrays.asList(expectedProcessors).toString());
+            }
+
+            @Override
+            protected void describeMismatchSafely(ProcessorsTrace item, Description description)
+            {
+                description.appendText("was ").appendValue(item.getExecutedProcessors());
```

Comment:
It is just so the message on failure is consistent with the rest of the messages generated by hamcrest. This template i got from BaseMatcher#describeMismatch()